src/test/scala/julienrf/json/derived/FieldAdapterSuite.scala

```diff
@@ -0,0 +1,169 @@
+            Gen.alphaStr.map(s => (Baz(s), Json.obj("s_c" -> s, "type" -> "Baz"))),
+            Gen.const((Bah, Json.obj("type" -> "Bah"))
+            )
+          ))
+
+
+      jsonIdentityLaw[Foo]
+    }
+
+    scenario("sum types with options") {
+      sealed trait Foo
+      case class Bar(xC: Int) extends Foo
+      case class Baz(sC: String) extends Foo
+      case object Bah extends Foo
+      case class Bat(oC: Option[String]) extends Foo
+
+      implicit val fooFormat: OFormat[Foo] = flat.oformat((__ \ "type").format[String], snakeAdapter())
+
+      implicit val fooArbitrary: Arbitrary[(Foo, JsValue)] =
+        Arbitrary(
+          Gen.oneOf(
+            arbitrary[Int].map(i => (Bar(i), Json.obj("x_c" -> i, "type" -> "Bar"))),
+            Gen.alphaStr.map(s => (Baz(s), Json.obj("s_c" -> s, "type" -> "Baz"))),
+            Gen.const((Bah, Json.obj("type" -> "Bah"))),
+            arbitrary[Option[String]].map(s => (Bat(s), Json.obj("type" -> "Bat") ++ s.fold(Json.obj())(x => Json.obj("o_c" -> x))))
+          ))
+
+      jsonIdentityLaw[Foo]
+    }
+
+
+
+    scenario("recursive type") {
+      sealed trait Tree
+      case class Leaf(lS: String) extends Tree
+      case class Node(lhsSnake: Tree, rhsSnake: Tree) extends Tree
+
+      def writeTree(tree: Tree): JsValue = tree match {
+        case n: Node =>
+          Json.obj(
+            "type" -> "Node",
+            "lhs_snake" -> writeTree(n.lhsSnake),
+            "rhs_snake" -> writeTree(n.rhsSnake)
+          )
+        case l: Leaf => Json.obj(
+          "type" -> "Leaf",
+          "l_s" -> l.lS
+        )
+      }
+
+      implicit val arbitraryTree: Arbitrary[Tree] = {
+        def atDepth(depth: Int): Gen[Tree] =
+          if (depth < 3) {
+            Gen.oneOf(
+              arbitrary[String].map(Leaf),
+              for {
+                lhs <- atDepth(depth + 1)
+                rhs <- atDepth(depth + 1)
+              } yield Node(lhs, rhs)
+            )
+          } else arbitrary[String].map(Leaf)
+        Arbitrary(atDepth(0))
+      }
+
+      implicit val arbitraryTreeWithJsValue: Arbitrary[(Tree, JsValue)] = {
+        Arbitrary(for (t <- arbitrary[Tree]) yield (t, writeTree(t)))
+      }
+
+      {
+        lazy val treeReads: Reads[Tree] = flat.reads[Tree]((__ \ "type").read[String], snakeAdapter(1))
+        lazy val treeWrites: OWrites[Tree] = flat.owrites((__ \ "type").write[String], snakeAdapter(1))
+        implicit lazy val treeFormat: OFormat[Tree] = OFormat.apply[Tree](treeReads, treeWrites)
+        jsonIdentityLaw[Tree]
+      }
+    }
+
+  }
+
+  def snakeAdapter(max:Int = 2) = new Adapter {
+    var nameMap = Map[String, Int]()
+
+    def increment(v1: String) = this.synchronized {
+      nameMap.get(v1).fold(nameMap += v1 -> 1)(i => nameMap += v1 -> (i + 1))
+      if (nameMap(v1) > max) throw new RuntimeException(s"Snake conversion applied more than $max times to field: $v1")
```

Comment:
This is to check the number of times the adapter function is called during a test. It should be called only once when deriving `Reads` and once when deriving `OWrites`